hs-tendermint/src/Network/Tendermint/Internal/RPCClient.hs

```diff
@@ -0,0 +1,114 @@
+module Network.Tendermint.Internal.RPCClient where
+
+import           Control.Applicative    ((<|>))
+import           Control.Exception      (Exception)
+import           Control.Monad.Catch    (MonadThrow, throwM)
+import           Control.Monad.IO.Class (MonadIO, liftIO)
+import           Control.Monad.Reader   (MonadReader, ask)
+import           Data.Aeson             (FromJSON (..), ToJSON (..), Value (..),
+                                         (.:), (.:?), (.=))
+import qualified Data.Aeson             as Aeson
+import qualified Data.ByteString        as B
+import qualified Data.ByteString.Lazy   as BL
+import           Data.Text              (Text, unpack)
+import qualified Network.HTTP.Simple    as HTTP
+import           System.Random          (randomIO)
+
+-- | JSON-RPC request.
+data Request = Request
+  { requestMethod :: !MethodName
+  , requestId     :: !Int
+  , requestParams :: !Value
+  }
+
+instance ToJSON Request where
+  toJSON (Request (MethodName method) rid params) = Aeson.object
+    [ "jsonrpc" .= String "2.0"
+    , "method"  .= method
+    , "params"  .= params
+    , "id"      .= rid
+    ]
+
+
+-- | JSON-RPC response.
+data Response = Response
+  { rsResult :: !(Either RpcError Value)
+  } deriving (Eq, Show)
+
+instance FromJSON Response where
+  parseJSON = Aeson.withObject "JSON-RPC response object" $ \v -> Response
+    <$> (Right <$> v .: "result" <|> Left <$> v .: "error")
+
+
+-- | JSON-RPC error message
+data RpcError = RpcError
+  { errCode    :: !Int
+  , errMessage :: !Text
+  , errData    :: !(Maybe Value)
+  } deriving Eq
+
+instance Show RpcError where
+  show (RpcError code msg dat) =
+      "JSON-RPC error " ++ show code ++ ": " ++ unpack msg
+        ++ ". Data: " ++ show dat
+
+instance FromJSON RpcError where
+  parseJSON = Aeson.withObject "JSON-RPC error object" $ \v -> RpcError
+    <$> v .: "code"
+    <*> v .: "message"
+    <*> v .:? "data"
+
+
+data JsonRpcException
+  = ParsingException String
+  | CallException RpcError
+  deriving (Eq, Show)
+
+instance Exception JsonRpcException
+
+
+-- | Name of called method.
+newtype MethodName = MethodName Text deriving (Eq, Show)
```

Comment:
if you derive `IsString` then you don't have to use the constructor in the RPC calls,

e.g. 

```
tx :: RequestTx -> TendermintM ResultTx
tx = RPC.remote "tx"
```

Reply:
if we add IsString then is there any advantage over just using `type MethodName = Text`?
I added the newtype so by looking at the type of remote it's clear what you need to pass.

Reply:
Yeah I’m this case there might be nothing to gain since probably every caller will use a string literal to name the method